File: frontend/src/api/inbox.ts
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query"
import { api } from "@/lib/api"

export type Ticket = {
  id: string
  user_id: string
  account_id: string
  topic_id: string | null
  linked_task_id: string | null
  thread_id: string
  subject: string
  from_name: string | null
  from_email: string
  status: "open" | "waiting" | "done" | "snoozed"
  priority: "high" | "med" | "low"
  ai_summary: string | null
  ai_intent: string | null
  due_at: string | null
  snoozed_until: string | null
  unread: boolean
  last_message_at: string
  created_at: string
  updated_at: string
  deleted_at: string | null
}

export type TicketMessage = {
  id: string
  ticket_id: string
  direction: "in" | "out"
  from_name: string | null
  from_email: string | null
  to_emails: string[] | null
  body_text: string | null
  body_html: string | null
  sent_at: string
  external_id: string | null
}

export type InboxRule = {
  id: string
  user_id: string
  name: string
  conditions: Record<string, unknown>
  actions: unknown[]
  position: number
  active: boolean
  run_count: number
  created_at: string
}

export function useTicketsQuery(params?: { status?: string; account_id?: string }) {
  return useQuery({
    queryKey: ["tickets", params],
    queryFn: async () => {
      const { data } = await api.get<Ticket[]>("/tickets", { params })
      return data
    },
  })
}

export function useTicketMessagesQuery(ticketId: string | null) {
  return useQuery({
    queryKey: ["ticket-messages", ticketId],
    queryFn: async () => {
      const { data } = await api.get<TicketMessage[]>(`/tickets/${ticketId}/messages`)
      return data
    },
    enabled: !!ticketId,
  })
}

export function useUpdateTicketMutation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, ...payload }: Partial<Ticket> & { id: string }) => {
      const { data } = await api.patch<Ticket>(`/tickets/${id}`, payload)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["tickets"] })
    },
  })
}

export function useInboxRulesQuery() {
  return useQuery({
    queryKey: ["inbox-rules"],
    queryFn: async () => {
      const { data } = await api.get<InboxRule[]>("/inbox/rules")
      return data
    },
  })
}

export function useCreateInboxRuleMutation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: Omit<InboxRule, "id" | "user_id" | "run_count" | "created_at">) => {
      const { data } = await api.post<InboxRule>("/inbox/rules", payload)
      return data
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["inbox-rules"] }),
  })
}

export function useUpdateInboxRuleMutation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, ...payload }: Partial<InboxRule> & { id: string }) => {
      const { data } = await api.patch<InboxRule>(`/inbox/rules/${id}`, payload)
      return data
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["inbox-rules"] }),
  })
}

export function useDeleteInboxRuleMutation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      await api.delete(`/inbox/rules/${id}`)
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["inbox-rules"] }),
  })
}

File: backend/app/schemas/inbox.py
from datetime import datetime
from enum import Enum
from typing import Any
from pydantic import BaseModel, Field


class TicketStatus(str, Enum):
    open = "open"
    waiting = "waiting"
    done = "done"
    snoozed = "snoozed"


class TicketPriority(str, Enum):
    high = "high"
    med = "med"
    low = "low"


# ─── EmailAccount ─────────────────────────────────────────

class EmailAccountCreate(BaseModel):
    provider: str = Field(..., max_length=20)
    email: str = Field(..., max_length=255)
    display_name: str | None = Field(default=None, max_length=255)
    imap_host: str | None = Field(default=None, max_length=255)
    imap_port: int | None = None


class EmailAccountResponse(BaseModel):
    id: str
    user_id: str
    provider: str
    email: str
    display_name: str | None = None
    imap_host: str | None = None
    imap_port: int | None = None
    active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ─── TicketMessage ────────────────────────────────────────

class TicketMessageResponse(BaseModel):
    id: str
    ticket_id: str
    direction: str
    from_name: str | None = None
    from_email: str | None = None
    to_emails: list[str] | None = None
    body_text: str | None = None
    body_html: str | None = None
    sent_at: datetime
    external_id: str | None = None

    model_config = {"from_attributes": True}


class TicketMessageCreate(BaseModel):
    body_text: str | None = None
    body_html: str | None = None
    to_emails: list[str] | None = None


# ─── Ticket ───────────────────────────────────────────────

class TicketResponse(BaseModel):
    id: str
    user_id: str
    account_id: str
    topic_id: str | None = None
    linked_task_id: str | None = None
    thread_id: str
    subject: str
    from_name: str | None = None
    from_email: str
    status: str
    priority: str
    ai_summary: str | None = None
    ai_intent: str | None = None
    due_at: datetime | None = None
    snoozed_until: datetime | None = None
    unread: bool
    last_message_at: datetime
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None = None

    model_config = {"from_attributes": True}


class TicketUpdate(BaseModel):
    status: TicketStatus | None = None
    priority: TicketPriority | None = None
    topic_id: str | None = None
    linked_task_id: str | None = None
    due_at: datetime | None = None
    snoozed_until: datetime | None = None
    unread: bool | None = None


# ─── InboxRule ────────────────────────────────────────────

class InboxRuleCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    conditions: dict[str, Any]
    actions: list[Any]
    position: int = 0
    active: bool = True


class InboxRuleUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    conditions: dict[str, Any] | None = None
    actions: list[Any] | None = None
    position: int | None = None
    active: bool | None = None


class InboxRuleResponse(BaseModel):
    id: str
    user_id: str
    name: str
    conditions: dict[str, Any]
    actions: list[Any]
    position: int
    active: bool
    run_count: int
    created_at: datetime

    model_config = {"from_attributes": True}


# ─── TicketActivity ───────────────────────────────────────

class TicketActivityResponse(BaseModel):
    id: str
    ticket_id: str
    kind: str
    actor: str | None = None
    payload: dict[str, Any] | None = None
    at: datetime

    model_config = {"from_attributes": True}
